File: rust-sdk/src/verify.rs
//! Independent receipt verifier.

use crate::canonicalize::canonicalize_bytes;
use crate::crypto::{sha256_hex, verify as verify_sig};
use crate::receipt::{SignedReceipt, GENESIS_HASH};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Debug, Default, Serialize, Deserialize)]
pub struct VerifyResult {
    pub valid: bool,
    pub canonical_hash_matches: bool,
    pub signature_valid: bool,
    pub chain_link_valid: Option<bool>,
    pub errors: Vec<String>,
}

pub fn verify_receipt(
    signed: &SignedReceipt,
    public_keys: &HashMap<String, String>,
    previous_receipt: Option<&SignedReceipt>,
) -> VerifyResult {
    let mut result = VerifyResult::default();

    // 0. structural guard: `integrity` must be present and an object.
    //    Mirror the hardened Go/Java behavior — bail out early as invalid.
    if signed
        .receipt
        .get("integrity")
        .and_then(Value::as_object)
        .is_none()
    {
        result
            .errors
            .push("integrity block missing or not an object".to_string());
        result.valid = false;
        return result;
    }

    // 1. recompute receipt_hash
    let mut body = signed.receipt.clone();
    if let Some(integrity) = body.get_mut("integrity").and_then(Value::as_object_mut) {
        integrity.insert("receipt_hash".to_string(), Value::String(String::new()));
    }
    let expected = sha256_hex(&canonicalize_bytes(&body));
    let got = signed
        .receipt
        .get("integrity")
        .and_then(|i| i.get("receipt_hash"))
        .and_then(|v| v.as_str())
        .unwrap_or("");
    if expected == got {
        result.canonical_hash_matches = true;
    } else {
        result.errors.push(format!(
            "canonical hash mismatch: expected {expected}, got {got}"
        ));
    }

    // 2. signatures
    let canon_sign = canonicalize_bytes(&signed.receipt);
    let mut any = false;
    for s in &signed.signatures {
        // Defense-in-depth: reject any signature whose alg is not exactly
        // "EdDSA" before running Ed25519 verification (spec §3).
        if s.alg != "EdDSA" {
            result
                .errors
                .push(format!("unsupported signature alg={} for kid={}", s.alg, s.kid));
            continue;
        }
        match public_keys.get(&s.kid) {
            Some(pk) => {
                if verify_sig(&canon_sign, &s.sig, pk) {
                    any = true;
                } else {
                    result.errors.push(format!("signature invalid for kid={}", s.kid));
                }
            }
            None => {
                result.errors.push(format!("no public key for kid={}", s.kid));
            }
        }
    }
    result.signature_valid = any;

    // 3. chain link
    let integrity = signed.receipt.get("integrity");
    // chain_height must be a positive integer (>= 1). A non-integer or a
    // non-integral JSON number leaves `height` as None -> reported invalid.
    let height: Option<i64> = integrity
        .and_then(|i| i.get("chain_height"))
        .and_then(Value::as_i64);
    let prev_hash_claim = integrity
        .and_then(|i| i.get("previous_receipt_hash"))
        .and_then(|v| v.as_str())
        .unwrap_or("");

    match height {
        Some(h) if h >= 1 => {
            if let Some(prev) = previous_receipt {
                let prev_hash = prev
                    .receipt
                    .get("integrity")
                    .and_then(|i| i.get("receipt_hash"))
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                let prev_height = prev
                    .receipt
                    .get("integrity")
                    .and_then(|i| i.get("chain_height"))
                    .and_then(Value::as_i64)
                    .unwrap_or(0);
                let link_ok = prev_hash == prev_hash_claim;
                let height_ok = h == prev_height + 1;
                result.chain_link_valid = Some(link_ok && height_ok);
                if !link_ok {
                    result.errors.push(format!(
                        "Chain link broken: previous_receipt_hash {prev_hash_claim} \
                         does not match previous receipt's receipt_hash {prev_hash}"
                    ));
                }
                if !height_ok {
                    result.errors.push(format!(
                        "Chain height not contiguous: expected {}, got {h}",
                        prev_height + 1
                    ));
                }
            } else if h == 1 || prev_hash_claim == GENESIS_HASH {
                // Genesis reference and chain_height 1 must agree.
                let genesis_ok = h == 1 && prev_hash_claim == GENESIS_HASH;
                result.chain_link_valid = Some(genesis_ok);
                if !genesis_ok {
                    result.errors.push(format!(
                        "Genesis inconsistency: chain_height {h} with \
                         previous_receipt_hash {prev_hash_claim} (chain_height 1 \
                         must reference GENESIS_HASH, and vice-versa)"
                    ));
                }
            }
            // else: mid-chain (height > 1) without predecessor -> leave
            // chain_link_valid as None: position not attested, but not failed.
        }
        _ => {
            // chain_height missing, non-integral, or < 1.
            result.chain_link_valid = Some(false);
            result.errors.push(format!(
                "Invalid chain_height: {}",
                integrity
                    .and_then(|i| i.get("chain_height"))
                    .map(|v| v.to_string())
                    .unwrap_or_else(|| "null".to_string())
            ));
        }
    }

    let chain_ok = result.chain_link_valid.unwrap_or(true);
    result.valid = result.canonical_hash_matches && result.signature_valid && chain_ok;
    result
}

File: python-sdk/src/askledger/receipts/verify.py
"""Independent receipt verifier. Compatible with TS-signed receipts."""

from __future__ import annotations

import copy
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .canonicalize import canonicalize_bytes
from .crypto import sha256_hex, verify as ed_verify
from .receipt import GENESIS_HASH


@dataclass
class VerifyResult:
    valid: bool = False
    canonical_hash_matches: bool = False
    signature_valid: bool = False
    chain_link_valid: Optional[bool] = None
    errors: List[str] = field(default_factory=list)


def verify_receipt(
    signed: Dict[str, Any],
    public_keys: Dict[str, str],
    previous_receipt: Optional[Dict[str, Any]] = None,
) -> VerifyResult:
    """Verify a signed receipt.

    Args:
      signed: the SignedReceipt envelope (`{receipt, signatures, ...}`)
      public_keys: map of kid -> base64 public key
      previous_receipt: optional prior SignedReceipt for chain-link check
    """
    result = VerifyResult()

    # 1. Recompute receipt_hash
    receipt = signed["receipt"]
    body = copy.deepcopy(receipt)
    body["integrity"]["receipt_hash"] = ""
    expected_hash = sha256_hex(canonicalize_bytes(body))
    if expected_hash == receipt["integrity"]["receipt_hash"]:
        result.canonical_hash_matches = True
    else:
        result.errors.append(
            f"Canonical hash mismatch: expected {expected_hash}, got "
            f"{receipt['integrity']['receipt_hash']}"
        )

    # 2. Verify at least one signature
    canon_signing = canonicalize_bytes(receipt)
    any_valid = False
    for sig in signed["signatures"]:
        kid = sig["kid"]
        # §3: reject any signature whose alg is not exactly "EdDSA" before
        # running Ed25519 verification (prevents algorithm confusion).
        alg = sig.get("alg")
        if alg != "EdDSA":
            result.errors.append(
                f"Unsupported signature alg={alg!r} for kid={kid} (expected 'EdDSA')"
            )
            continue
        pk = public_keys.get(kid)
        if not pk:
            result.errors.append(f"No public key supplied for kid={kid}")
            continue
        if ed_verify(canon_signing, sig["sig"], pk):
            any_valid = True
        else:
            result.errors.append(f"Signature invalid for kid={kid}")
    result.signature_valid = any_valid

    # 3. Chain link
    integrity = receipt["integrity"]
    height = integrity.get("chain_height")
    prev_hash_claim = integrity.get("previous_receipt_hash")

    if not isinstance(height, int) or isinstance(height, bool) or height < 1:
        result.chain_link_valid = False
        result.errors.append(f"Invalid chain_height: {height}")
    elif previous_receipt is not None:
        prev = previous_receipt["receipt"]["integrity"]
        link_ok = prev["receipt_hash"] == prev_hash_claim
        height_ok = height == prev["chain_height"] + 1
        result.chain_link_valid = link_ok and height_ok
        if not link_ok:
            result.errors.append(
                f"Chain link broken: previous_receipt_hash "
                f"{prev_hash_claim} does not match "
                f"prev receipt's receipt_hash {prev['receipt_hash']}"
            )
        if not height_ok:
            result.errors.append(
                f"Chain height not contiguous: expected "
                f"{prev['chain_height'] + 1}, got {height}"
            )
    elif height == 1 or prev_hash_claim == GENESIS_HASH:
        # Genesis reference and chain_height 1 must agree with each other.
        genesis_ok = height == 1 and prev_hash_claim == GENESIS_HASH
        result.chain_link_valid = genesis_ok
        if not genesis_ok:
            result.errors.append(
                f"Genesis inconsistency: chain_height {height} with "
                f"previous_receipt_hash {prev_hash_claim} (chain_height 1 must "
                f"reference GENESIS_HASH, and vice-versa)"
            )
    # else: mid-chain (height > 1) without predecessor -> chain_link_valid
    # stays None (unset): position not attested, but not failed.

    result.valid = (
        result.canonical_hash_matches
        and result.signature_valid
        and (result.chain_link_valid is not False)
    )
    return result
